packages/starlight/utils/plugins.ts

```diff
@@ -0,0 +1,143 @@
+import type { AstroIntegration, AstroIntegrationLogger } from 'astro';
+import { z } from 'astro/zod';
+import type { StarlightUserConfig } from '../utils/user-config';
+import { errorMap } from '../utils/error-map';
+
+/**
+ * Runs Starlight plugins in the order that they are configured after validating the plugins config
+ * and returns the final user config that may have been updated by the plugins and a list of any
+ * integrations added by the plugins.
+ * Note that the returned config is not validated and should still be validated by the caller.
+ */
+export async function runPlugins(
+	{ plugins, ...config }: StarlightUserConfigWithPlugins,
+	logger: AstroIntegrationLogger
+) {
+	const parsedPluginsConfig = starlightPluginsConfigSchema.safeParse({ plugins }, { errorMap });
+
+	if (!parsedPluginsConfig.success) {
+		throw new Error(
+			'Invalid plugins config passed to starlight integration\n' +
+				parsedPluginsConfig.error.issues.map((i) => i.message).join('\n')
+		);
+	}
+
+	let userConfig = config;
+	const integrations: AstroIntegration[] = [];
+
+	for (const { name, plugin } of parsedPluginsConfig.data.plugins) {
+		await plugin({
+			config: userConfig,
+			logger: logger.fork(name),
+			updateConfig(newConfig) {
+				userConfig = { ...userConfig, ...newConfig };
+			},
+			addIntegration(integration) {
+				integrations.push(integration);
+			},
+		});
+	}
+
+	return { integrations, userConfig };
+}
+
+// https://github.com/withastro/astro/blob/144815c6ff295fe54eef3c641ad929e3c7b2b2fd/packages/astro/src/core/config/schema.ts#L101
+const astroIntegrationSchema = z.object({
+	name: z.string(),
+	hooks: z.object({}).passthrough().default({}),
+}) as z.Schema<AstroIntegration>;
+
+/**
+ * A plugin `config` and `updateConfig` argument are purposely not validated using the Starlight
+ * user config schema but properly typed for user convenience because we do not want to run any of
+ * the Zod `transform`s used in the user config schema when running plugins.
+ * The final user config should be validated by the caller after running all the plugins.
+ */
+const starlightPluginSchema = z.object({
+	/** Name of the Starlight plugin. */
+	name: z.string(),
+	/**
+	 * Plugin function called with an object containing various values that can be used by the
+	 * plugin to interact with Starlight.
+	 */
+	plugin: z.function(
```

Comment:
Would it make sense to also pass some [data from the `astro:config:setup`](https://docs.astro.build/en/reference/integrations-reference/#astroconfigsetup) hook here? I think it’s OK to require adding an integration to use any of the functions, but it would probably be helpful for plugins to access [`config`](https://docs.astro.build/en/reference/integrations-reference/#config-option), [`command`](https://docs.astro.build/en/reference/integrations-reference/#command-option), and [`isRestart`](https://docs.astro.build/en/reference/integrations-reference/#isrestart-option). 

Example use cases:
- check if an integration is already in the Astro config before adding it
- only run some initialisation code on dev/build start, but not when the dev server restarts

Reply:
Yeah, because otherwise you could be in the weird situation of using `addIntegration()` to add an integration that reads `config` and _then_ itself adds an integration based on that, which is pretty clunky. (The use case I have in mind is the Algolia plugin, which could add Preact or React depending on what a user is already using.)